collect_app/src/main/java/org/odk/collect/android/application/initialization/SavepointsImporter.kt

```diff
@@ -0,0 +1,75 @@
+    private val instancesRepositoryProvider: InstancesRepositoryProvider,
+    private val formsRepositoryProvider: FormsRepositoryProvider,
+    private val savepointsRepositoryProvider: SavepointsRepositoryProvider,
+    private val storagePathProvider: StoragePathProvider
+) : Upgrade {
+    override fun key(): String {
+        return MetaKeys.OLD_SAVEPOINTS_IMPORTED
+    }
+
+    override fun run() {
+        projectsRepository.getAll().forEach { project ->
+            val instancesRepository = instancesRepositoryProvider.get(project.uuid)
+            val formsRepository = formsRepositoryProvider.get(project.uuid)
+            val savepointsRepository = savepointsRepositoryProvider.get(project.uuid)
+            val cacheDir = File(storagePathProvider.getOdkDirPath(StorageSubdirectory.CACHE, project.uuid))
+            val instancesDir = File(storagePathProvider.getOdkDirPath(StorageSubdirectory.INSTANCES, project.uuid))
+
+            importSavepointsThatBelongToSavedForms(instancesRepository, formsRepository, savepointsRepository, cacheDir)
+
+            importSavepointsThatBelongToBlankForms(formsRepository, savepointsRepository, cacheDir, instancesDir)
+        }
+    }
+
+    private fun importSavepointsThatBelongToSavedForms(instancesRepository: InstancesRepository, formsRepository: FormsRepository, savepointsRepository: SavepointsRepository, cacheDir: File) {
+        instancesRepository.all.forEach { instance ->
+            if (instance.deletedDate == null) {
+                val savepointFile = File(cacheDir, File(instance.instanceFilePath).name.plus(".save"))
+                if (savepointFile.exists() && savepointFile.lastModified() > instance.lastStatusChangeDate) {
+                    formsRepository.getAllByFormIdAndVersion(instance.formId, instance.formVersion).firstOrNull()?.let { form ->
+                        if (!form.isDeleted) {
+                            savepointsRepository.save(Savepoint(form.dbId, instance.dbId, savepointFile.absolutePath, instance.instanceFilePath))
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun importSavepointsThatBelongToBlankForms(formsRepository: FormsRepository, savepointsRepository: SavepointsRepository, cacheDir: File, instancesDir: File) {
+        formsRepository.all.sortedByDescending { form -> form.date }.forEach { form ->
+            if (!form.isDeleted) {
+                val formFileName = File(form.formFilePath).name.substringBeforeLast(".xml")
+
+                cacheDir.listFiles { file ->
+                    file.name.startsWith("${formFileName}_") && file.name.endsWith(".xml.save")
+                }?.forEach { savepointFile ->
+                    if (savepointFile.lastModified() > form.date) {
+                        val alreadyUsed = savepointsRepository.getAll().any { savepoint -> savepoint.savepointFilePath == savepointFile.absolutePath }
+                        if (!alreadyUsed) {
```

Comment:
Is this possible? I don't see a test for this case.

Reply:
This test will fail `ifThereAreMultipleVersionsOfTheSameBlankFormWithSavepoints_allSavepointsShouldBeImported` and it's related to what I said here https://github.com/getodk/collect/pull/6068#discussion_r1558454310
Imaging you have two versions of the same form:
`form.xml` (v1) with a savepoint `form_2024-04-10_01-35-41.xml.save`
`form_1.xml` (v2) with a savepoint `form_1_2024-04-10_01-45-41.xml.save`

during importing it will look like:
1. We take the newer form into account first `form_1.xml` and look for savepoints with names that meet the requirements  (the file name starts with `form_1_ `and ends with `.xml.save`). There is only one file like that so we import it to the database.
2. Then we take the old version of that form `form.xml` and try to find a savepoint in the same way. In this case, both files will meet the requirements so depending on which was first on the list that one will be loaded and it might be a savepoint created for the newer version `form_1_2024-04-10_01-45-41.xml.save`.

That's why I need to load savepoints for blank forms starting from the newest forms and additionally checking if a given savepoint is not imported already.

Reply:
There's something strange happening: if I comment on this `if` and run the tests, that one sometimes fails and sometimes passes.


<img width="2416" alt="Screenshot 2024-05-13 at 13 48 53" src="https://github.com/getodk/collect/assets/556280/e06da55d-3cbc-4aa2-bd0d-d0bd33467788">

Reply:
No, it's completely ok. The test that is supposed to fail when you remove that if statement does not always fail because the order of files when you read them from cache is random: https://github.com/getodk/collect/pull/6068/files#diff-a00717447283323def9e47f289dc2227bc7bd76495ff73791e1fb8d6b8b54748R56 when the one that should be used is first then you save it and despite the fact you try to save the second one it will pass because you can't save another savepoint with the same form and instance ids https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/database/savepoints/DatabaseSavepointsRepository.kt#L56
If the order changes and the first one is the one that should be ignored (alreadyused) then it fails.

Reply:
Ah because `listFiles` isn't ordered? We definitely want to avoid non deterministic tests at all costs. The fact that this test will sometimes pass when the code is in a failing state makes it fairly hard to reason about. A common trick here would be to use dependency inversion to extract the "random" element (passing in the files for example) or sort in the code (although this removes control from the test).

That aside, I think there's potentially a different test approach to this problem that also demonstrates the current code needs to be amended (although I might be misunderstanding). Should this following test not pass?

```kotlin
val blankForm = createBlankForm(project, "sampleForm", "1", "1", date = 1)
val savePointFile = createSavepointFile(project, "sampleForm_1_${System.currentTimeMillis()}.xml")

savepointsImporter.run()

val savepoints = projectDependencyProvider.savepointsRepository.getAll()
assertThat(savepoints, empty())
```

My understanding is that the save point files always have a name structured like `<form name>_<timestamp>.xml.save`, so a save point named `<form name>_1_<timestamp>.xml.save` should not be "matched".

Reply:
> That aside, I think there's potentially a different test approach to this problem that also demonstrates the current code needs to be amended (although I might be misunderstanding). Should this following test not pass?

Such a test would pass because we don't have enough information what I have described here https://github.com/getodk/collect/pull/6068#discussion_r1558454310. The problem is that for blank forms we don't know the timestamp part of a savepoint file name and we need to look for files with names starting with our form name.

In your example I need to look for files that start with `sampleForm_` and `sampleForm_1_${System.currentTimeMillis()}.xml` is matched.

To solve the problem with the random order of savepoints retrieved from cache I can just add sorting:

```java
                cacheDir.listFiles { file ->
                    file.name.startsWith("${formFileName}_") && file.name.endsWith(".xml.save")
                }?.sortedByDescending { it.name }?.forEach { savepointFile ->
                    if (savepointFile.lastModified() > form.date) {
                        val alreadyUsed = savepointsRepository.getAll().any { savepoint -> savepoint.savepointFilePath == savepointFile.absolutePath }
                        if (!alreadyUsed) {
                            savepointsRepository.save(Savepoint(form.dbId, null, savepointFile.absolutePath, File(instancesDir, "$formFileName/$formFileName.xml").absolutePath))
                        }
                    }
                }
```
Then the test will always fail if yu comment out `if (!alreadyUsed) {`. What do you think?

Reply:
> Such a test would pass because we don't have enough information what I have described here https://github.com/getodk/collect/pull/6068#discussion_r1558454310. 

If you copy that code in as a new test it fails though. 

> In your example I need to look for files that start with sampleForm_ and sampleForm_1_${System.currentTimeMillis()}.xml is matched.

I could definitely be getting confused here (you're making all these improvements **because of** just how confusing it was after all), but my understanding is that it shouldn't be matched as we don't want to import a save point for a different form version. If old save point files are `<form file name>_<timestamp>.xml.save`, then surely I only want to import files that exactly match that format.

Reply:
> then surely I only want to import files that exactly match that format

This is not possible because, in the case of savepoints that belong to blank forms, you don't know the timestamp part of a file name. 
To avoid such issues I start importing forms from the newer ones: https://github.com/getodk/collect/pull/6068/files#diff-a00717447283323def9e47f289dc2227bc7bd76495ff73791e1fb8d6b8b54748R52 and detect already used savepoints https://github.com/getodk/collect/pull/6068/files#diff-a00717447283323def9e47f289dc2227bc7bd76495ff73791e1fb8d6b8b54748R61
thanks to that if you have `sampleForm`  with sampleForm_1715764796206.xml.save savepoint and `sampleForm_1` with `sampleForm_1_1715764796207.xml.save`
once we do the job for `sampleForm_1` first, when we try to do the same for `sampleForm` even though `sampleForm_1_1715764796207.xml.save` is a match it won't be used thanks to https://github.com/getodk/collect/pull/6068/files#diff-a00717447283323def9e47f289dc2227bc7bd76495ff73791e1fb8d6b8b54748R61

but you are right this might be a problem when there is a savepoint for `sampleForm_1` but the form does not exist then the only form we want to import a savepoint for is `sampleForm` and if the first savepoint file taken into account is `sampleForm_1_1715764796207.xml.save` it will be used.

> Is the "usually" there a reference to some fuzziness I'm missing?

I said usually because I imagine people do not change form names when adding changes and bumping the version number and then forms are named with that _1, _2 etc. suffix. If a newer version of a form has a different name then it would be even better because savepoints would have different and easy-to-detect names as well.

Reply:
> This is not possible because, in the case of savepoints that belong to blank forms, you don't know the timestamp part of a file name.

Right, but can't we extract the "form name" part of the save point file name by just extracting everything before `_<timestamp>` without knowing it using a regex (or a similar fuzzy matching approach)? Then we could match save points to forms exactly. 

If I have a form with a filename `form.xml` then I'm looking for a save point named `form_<timestamp>.xml.save`. I don't need to know the exact timestamp to find that file, I just need to be able to extract everything before a regex matching `_<timestamp>.xml.save`. The examples we've been talking about use different timestamp formats, but if it's just millis then it'd be something like `_\d*.xml.save` and if it's a formatted timestamp (like `2024-04-10_01-35-41`) then it can be even more specific. I'm not sure if both are possible or just one of them.

Reply:
The timestamp is formated so it's like `2024-04-10_01-35-41` so we can try to substring after second to last underscore maybe? What do you think about this approach https://github.com/getodk/collect/pull/6068/commits/c910b7328aa1d0524f5861c7a0913d6de81e76f8?

Reply:
This really feels like a good place to use a regular expression (even though I'm useless at them). Playing around with [regex101](https://regex101.com), it seems like we can match the timestamp format with `\d{4}-\d{2}-\d{2}_\d{2}-\d{2}-\d{2}`. As far as I'm aware, Kotlin doesn't let us do "substring before" with a regex, but we can just cheat and split on it:

```kotlin
val timestampRegex = """\d{4}-\d{2}-\d{2}_\d{2}-\d{2}-\d{2}""".toRegex()
val formName = fileName.split(timestampRegex)[0]
```

We probably want to use capture groups (`()`) instead thought so we can ensure we only match on files that are actually save points (and we should have a test for that as well). Again playing around with regex101, I think that'd be something like:

```kotlin
val savepointFileRegex = """(.*)(\d{4}-\d{2}-\d{2}_\d{2}-\d{2}-\d{2})(.xml.save)""".toRegex()
val match = savePointFileRegex.matchEntire(fileName)
if (match != null) {
    val (formName, _, _) = match.destructured
    // Import save point file
}
```

Where the three groups are: any number of characters, `_` followed by a timestamp, literally `.xml.save`. I guess it could actually just be two groups, but that's how my brain was working!

Reply:
Ok now it uses regex. Please review. The tests are failing because of https://github.com/getodk/collect/pull/6138